src/healthcheck.c

```diff
@@ -0,0 +1,422 @@
+			nwarn("Failed to allocate memory for runtime command");
+			_exit(127);
+		}
+
+		runtime_argv[0] = (char *)runtime_path; /* Runtime executable */
+		runtime_argv[1] = "exec";		/* Runtime subcommand */
+		runtime_argv[2] = (char *)container_id; /* Container ID */
+
+		/* Copy healthcheck command and arguments */
+		for (int i = 0; i < argc; i++) {
+			runtime_argv[3 + i] = config->test[i];
+		}
+		runtime_argv[3 + argc] = NULL; /* NULL terminator */
+
+		/* Execute the runtime command */
+		if (execvp(runtime_path, runtime_argv) == -1) {
+			/* If execvp fails, exit with error code */
+			_exit(127); /* Command not found */
+		}
+	} else {
+		/* Parent process - wait for child to complete */
+		close(stderr_pipe[1]); /* Close write end of stderr pipe */
+		int status;
+		pid_t wait_result = waitpid(pid, &status, 0);
+
+		if (wait_result == -1) {
+			nwarnf("Failed to wait for healthcheck command: %s", strerror(errno));
+			close(stderr_pipe[0]);
+			return false;
+		}
+
+		/* Read stderr output */
+		char stderr_buffer[4096];
+		ssize_t stderr_len = read(stderr_pipe[0], stderr_buffer, sizeof(stderr_buffer) - 1);
+		close(stderr_pipe[0]);
+
+		if (stderr_len > 0) {
+			stderr_buffer[stderr_len] = '\0';
+			/* Trim trailing newlines */
+			while (stderr_len > 0 && (stderr_buffer[stderr_len - 1] == '\n' || stderr_buffer[stderr_len - 1] == '\r')) {
+				stderr_buffer[--stderr_len] = '\0';
+			}
+		} else {
+			stderr_buffer[0] = '\0';
+		}
+
+		if (WIFEXITED(status)) {
+			*exit_code = WEXITSTATUS(status);
+			if (*exit_code != 0) {
+				nwarnf("Healthcheck command failed (exit code %d): %s", *exit_code, config->test[0]);
+				if (stderr_len > 0) {
+					nwarnf("Healthcheck command stderr: %s", stderr_buffer);
+				}
+			}
+			return true;
+		} else if (WIFSIGNALED(status)) {
+			nwarnf("Healthcheck command terminated by signal %d: %s", WTERMSIG(status), config->test[0]);
+			if (stderr_len > 0) {
+				nwarnf("Healthcheck command stderr: %s", stderr_buffer);
+			}
+			*exit_code = 128 + WTERMSIG(status); /* Standard convention for signal termination */
+			return true;
+		} else {
+			nwarnf("Healthcheck command did not terminate normally: %s", config->test[0]);
+			if (stderr_len > 0) {
+				nwarnf("Healthcheck command stderr: %s", stderr_buffer);
+			}
+			*exit_code = -1;
+			return false;
+		}
+	}
+
+	/* This should never be reached */
+	return false;
+}
+
+/* Convert healthcheck status to string */
+const char *healthcheck_status_to_string(int status)
+{
+	if (status >= 0 && status < 4) {
+		return healthcheck_status_strings[status];
+	}
+	return "unknown";
+}
+
+/* Send healthcheck status update to Podman */
+bool healthcheck_send_status_update(const char *container_id, int status, int exit_code)
+{
+	if (container_id == NULL) {
+		return false;
+	}
+
+	const char *status_str = healthcheck_status_to_string(status);
+
+	/* Create simple JSON message for status update */
+	char json_msg[1024];
+	snprintf(json_msg, sizeof(json_msg),
```

Comment:
I'd remove this JSON part and use just minimalistic conmon protocol - JSON is over-complication here and poses injection risks.

Something like:
```
/* Include exit code in message for debugging when status indicates failure */
char message[256];
if (exit_code != 0 && status == HEALTHCHECK_UNHEALTHY) {
      snprintf(message, sizeof(message), "%s (exit_code: %d)", status_str, exit_code);
      write_or_close_sync_fd(&sync_pipe_fd, status, message);
 } else {
       write_or_close_sync_fd(&sync_pipe_fd, status, status_str);
 }
```